src/coreclr/jit/importer.cpp

```diff
@@ -11877,29 +11877,85 @@ void Compiler::impImportBlockCode(BasicBlock* block)
 
         // OSR is not yet supported for methods with explicit tail calls.
         //
-        // But we also do not have to switch these methods to be optimized as we should be
+        // But we also do not have to switch these methods to be optimized, as we should be
         // able to avoid getting trapped in Tier0 code by normal call counting.
         // So instead, just suppress adding patchpoints.
         //
         if (!compTailPrefixSeen)
         {
-            // The normaly policy is only to add patchpoints to the targets of lexically
-            // backwards branches.
+            // We only need to add patchpoints if the method can loop.
             //
             if (compHasBackwardJump)
             {
                 assert(compCanHavePatchpoints());
 
-                // Is the start of this block a suitable patchpoint?
+                // Is the start of this block a suitable patchpoint? We can only place patchpoints
+                // at stack empty points (so the only live state is the IL visible state).
                 //
-                if (((block->bbFlags & BBF_BACKWARD_JUMP_TARGET) != 0) && (verCurrentState.esStackDepth == 0))
+                if (verCurrentState.esStackDepth == 0)
                 {
-                    // We should have noted this earlier and bailed out of OSR.
+                    // By default we use the "source" strategy.
                     //
-                    assert(!block->hasHndIndex());
+                    const int patchpointStrategy = JitConfig.TC_PatchpointStrategy();
```

Comment:
Would it be clearer to have `enum PatchpointStrategy { BackedgeSource, BackedgeTarget, Adaptive };` and return enum value from `TC_PatchpointStrategy()`?

Reply:
The config system is kind of primitive, so probably not.